File: heartbeat-system/core/heartbeat.py
"""
Heartbeat System - 心跳系统
定时执行维护任务：状态同步、情感衰减、垃圾清理
"""

import os
import sys
import time
import json
import logging
from datetime import datetime
from pathlib import Path

# 配置日志
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s [heartbeat] %(levelname)s: %(message)s'
)
logger = logging.getLogger(__name__)


class HeartbeatConfig:
    """心跳配置"""
    
    # 心跳间隔（秒）
    PULSE_INTERVAL = 5        # 脉冲层：5秒
    RHYTHM_INTERVAL = 60      # 节律层：1分钟
    CYCLE_INTERVAL = 3600    # 周期层：1小时
    
    # 存储路径
    WORKSPACE_DIR = os.path.expanduser("~/.openclaw/workspace")
    MEMORY_DIR = os.path.join(WORKSPACE_DIR, "memory")
    EMOTION_DIR = os.path.join(MEMORY_DIR, "emotions")
    STATE_FILE = os.path.join(WORKSPACE_DIR, "heartbeat_state.json")
    
    # 任务开关
    TASKS = {
        "pulse": True,       # 存活确认
        "rhythm": True,      # 情感衰减、状态保存
        "cycle": True,      # 记忆整理、深度清理
    }


class HeartbeatTask:
    """心跳任务基类"""
    
    name = "base"
    
    def should_run(self, elapsed: int, last_run: int) -> bool:
        """判断是否应该执行"""
        return elapsed >= last_run
    
    def run(self) -> bool:
        """执行任务，返回是否成功"""
        raise NotImplementedError


class PulseTask(HeartbeatTask):
    """脉冲层任务：存活确认"""
    
    name = "pulse"
    interval = HeartbeatConfig.PULSE_INTERVAL
    
    def run(self) -> bool:
        # 简单确认存活
        return True


class RhythmTask(HeartbeatTask):
    """节律层任务：情感衰减、状态保存"""
    
    name = "rhythm"
    interval = HeartbeatConfig.RHYTHM_INTERVAL
    
    def run(self) -> bool:
        try:
            # 调用情感系统衰减
            self._emotion_decay()
            # 更新最后运行时间
            self._update_state()
            return True
        except Exception as e:
            logger.error(f"Rhythm task failed: {e}")
            return False
    
    def _emotion_decay(self):
        """情感衰减"""
        emotion_script = os.path.join(
            HeartbeatConfig.WORKSPACE_DIR,
            "templates/emotion-system/main.py"
        )
        if os.path.exists(emotion_script):
            os.system(f'python3 "{emotion_script}" decay > /dev/null 2>&1')
    
    def _update_state(self):
        """更新状态"""
        state = self._load_state()
        state['last_rhythm'] = datetime.now().isoformat()
        self._save_state(state)
    
    def _load_state(self) -> dict:
        if os.path.exists(HeartbeatConfig.STATE_FILE):
            with open(HeartbeatConfig.STATE_FILE, 'r') as f:
                return json.load(f)
        return {}
    
    def _save_state(self, state: dict):
        with open(HeartbeatConfig.STATE_FILE, 'w') as f:
            json.dump(state, f, indent=2)


class CycleTask(HeartbeatTask):
    """周期层任务：记忆整理、深度清理"""
    
    name = "cycle"
    interval = HeartbeatConfig.CYCLE_INTERVAL
    
    def run(self) -> bool:
        try:
            self._cleanup_old_files()
            self._update_state()
            return True
        except Exception as e:
            logger.error(f"Cycle task failed: {e}")
            return False
    
    def _cleanup_old_files(self):
        """清理旧文件"""
        # 清理临时文件
        temp_patterns = ['*.tmp', '*.cache', '__pycache__']
        
        # 清理 memory/archive 中的旧记忆（超过90天）
        archive_dir = os.path.join(HeartbeatConfig.MEMORY_DIR, "archive")
        if os.path.exists(archive_dir):
            # 这里可以添加更复杂的清理逻辑
            pass
    
    def _update_state(self):
        state = self._load_state()
        state['last_cycle'] = datetime.now().isoformat()
        self._save_state(state)
    
    def _load_state(self) -> dict:
        if os.path.exists(HeartbeatConfig.STATE_FILE):
            with open(HeartbeatConfig.STATE_FILE, 'r') as f:
                return json.load(f)
        return {}
    
    def _save_state(self, state: dict):
        with open(HeartbeatConfig.STATE_FILE, 'w') as f:
            json.dump(state, f, indent=2)


class HeartbeatSystem:
    """心跳系统主控制器"""
    
    def __init__(self):
        self.config = HeartbeatConfig()
        self.tasks = {
            'pulse': PulseTask(),
            'rhythm': RhythmTask(),
            'cycle': CycleTask(),
        }
        
        self.state = self._load_state()
        self.running = False
        
        # 记录各层最后运行时间
        self.last_run = {
            'pulse': 0,
            'rhythm': 0,
            'cycle': 0,
        }
    
    def _load_state(self) -> dict:
        if os.path.exists(self.config.STATE_FILE):
            with open(self.config.STATE_FILE, 'r') as f:
                return json.load(f)
        return {
            'start_time': datetime.now().isoformat(),
            'last_pulse': None,
            'last_rhythm': None,
            'last_cycle': None,
        }
    
    def _save_state(self):
        with open(self.config.STATE_FILE, 'w') as f:
            json.dump(self.state, f, indent=2)
    
    def _update_last_run(self, task_name: str):
        self.last_run[task_name] = time.time()
        self.state[f'last_{task_name}'] = datetime.now().isoformat()
        self._save_state()
    
    def tick(self):
        """执行一次心跳"""
        elapsed = time.time()
        
        for task_name, task in self.tasks.items():
            interval = getattr(task, 'interval', 5)
            
            if elapsed - self.last_run.get(task_name, 0) >= interval:
                if task.run():
                    self._update_last_run(task_name)
                    logger.info(f"{task_name} task completed")
    
    def start(self, max_iterations: int = None):
        """启动心跳循环"""
        self.running = True
        iteration = 0
        
        logger.info("Heartbeat system started")
        
        while self.running:
            self.tick()
            
            iteration += 1
            if max_iterations and iteration >= max_iterations:
                break
            
            time.sleep(self.config.PULSE_INTERVAL)
        
        logger.info("Heartbeat system stopped")
    
    def stop(self):
        """停止心跳"""
        self.running = False
    
    def get_status(self) -> dict:
        """获取状态"""
        return {
            'running': self.running,
            'last_run': self.last_run,
            'state': self.state,
        }


def run_once():
    """运行一次心跳（用于测试）"""
    hb = HeartbeatSystem()
    hb.tick()
    print(json.dumps(hb.get_status(), indent=2))


def run_daemon():
    """作为守护进程运行"""
    hb = HeartbeatSystem()
    hb.start()


if __name__ == "__main__":
    import argparse
    
    parser = argparse.ArgumentParser(description='Heartbeat System')
    parser.add_argument('--once', action='store_true', help='运行一次')
    parser.add_argument('--daemon', action='store_true', help='守护进程模式')
    
    args = parser.parse_args()
    
    if args.once:
        run_once()
    elif args.daemon:
        run_daemon()
    else:
        run_once()

File: src/systems/heartbeat.ts
/**
 * Heartbeat System - 完善版心跳系统 (TypeScript)
 * 
 * 改进：
 * - 可插拔任务架构（注册自定义任务）
 * - 任务依赖管理
 * - 与情感/记忆系统联动
 * - 更详细的统计和日志
 */

import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';

// ============================================================================
// 类型定义
// ============================================================================

export interface HeartbeatConfig {
  pulseInterval: number;      // 脉冲层：5秒
  rhythmInterval: number;     // 节律层：60秒
  cycleInterval: number;     // 周期层：1小时
  workspaceDir: string;
  stateFile: string;
}

export interface HeartbeatState {
  startTime: string;
  lastPulse: string | null;
  lastRhythm: string | null;
  lastCycle: string | null;
  totalPulses: number;
  totalRhythms: number;
  totalCycles: number;
  failedTasks: number;
}

export interface TaskResult {
  success: boolean;
  taskName: string;
  message?: string;
  duration?: number;
  error?: string;
}

export interface Task {
  name: string;
  layer: 'pulse' | 'rhythm' | 'cycle';
  interval: number;  // 毫秒
  run: () => Promise<TaskResult>;
  enabled?: boolean;
}

export interface TaskStats {
  name: string;
  lastRun: number | null;
  runCount: number;
  successCount: number;
  failedCount: number;
  avgDuration: number;
  lastResult?: TaskResult;
}

// ============================================================================
// 配置
// ============================================================================

const DEFAULT_CONFIG: HeartbeatConfig = {
  pulseInterval: 5 * 1000,        // 5秒
  rhythmInterval: 60 * 1000,     // 1分钟
  cycleInterval: 3600 * 1000,    // 1小时
  workspaceDir: '',
  stateFile: '',
};

function initConfig(workspaceDir?: string): HeartbeatConfig {
  const wsDir = workspaceDir || path.join(os.homedir(), '.openclaw', 'workspace');
  return {
    ...DEFAULT_CONFIG,
    workspaceDir: wsDir,
    stateFile: path.join(wsDir, 'heartbeat_state.json'),
  };
}

// ============================================================================
// 任务基类与注册表
// ============================================================================

class TaskRegistry {
  private tasks: Map<string, Task> = new Map();
  private stats: Map<string, TaskStats> = new Map();
  
  register(task: Task): void {
    this.tasks.set(task.name, { ...task, enabled: task.enabled !== false });
    this.stats.set(task.name, {
      name: task.name,
      lastRun: null,
      runCount: 0,
      successCount: 0,
      failedCount: 0,
      avgDuration: 0,
    });
  }
  
  unregister(name: string): void {
    this.tasks.delete(name);
    this.stats.delete(name);
  }
  
  getTasks(layer?: string): Task[] {
    if (!layer) return Array.from(this.tasks.values());
    return Array.from(this.tasks.values()).filter(t => t.layer === layer);
  }
  
  getStats(name: string): TaskStats | undefined {
    return this.stats.get(name);
  }
  
  getAllStats(): TaskStats[] {
    return Array.from(this.stats.values());
  }
  
  updateStats(result: TaskResult, duration: number): void {
    const stats = this.stats.get(result.taskName);
    if (!stats) return;
    
    stats.lastRun = Date.now();
    stats.runCount++;
    if (result.success) {
      stats.successCount++;
    } else {
      stats.failedCount++;
    }
    // 移动平均
    stats.avgDuration = (stats.avgDuration * (stats.runCount - 1) + duration) / stats.runCount;
    stats.lastResult = result;
  }
}

// ============================================================================
// 内置任务
// ============================================================================

/**
 * 脉冲任务：存活确认 + 健康检查
 */
function createPulseTask(): Task {
  return {
    name: 'health_check',
    layer: 'pulse',
    interval: 5000,
    run: async () => {
      const start = Date.now();
      // 简单健康检查
      return {
        success: true,
        taskName: 'health_check',
        duration: Date.now() - start,
      };
    },
  };
}

/**
 * 节律任务：情感衰减 + 状态同步
 */
function createRhythmTask(workspaceDir: string): Task {
  return {
    name: 'emotion_decay',
    layer: 'rhythm',
    interval: 60000,
    run: async () => {
      const start = Date.now();
      const emotionStateFile = path.join(workspaceDir, 'memory', 'emotions', 'state.json');
      
      if (!fs.existsSync(emotionStateFile)) {
        return {
          success: true,
          taskName: 'emotion_decay',
          duration: Date.now() - start,
          message: 'No emotion state file',
        };
      }
      
      try {
        const state = JSON.parse(fs.readFileSync(emotionStateFile, 'utf-8'));
        
        // 应用衰减
        if (state.energy > 0.2) {
          state.energy = Math.max(0.2, state.energy - 0.01);
        }
        if (state.connection > 0) {
          state.connection = Math.max(0, state.connection - 0.005);
        }
        state.lastUpdate = new Date().toISOString();
        
        fs.writeFileSync(emotionStateFile, JSON.stringify(state, null, 2));
        
        return {
          success: true,
          taskName: 'emotion_decay',
          duration: Date.now() - start,
        };
      } catch (error) {
        return {
          success: false,
          taskName: 'emotion_decay',
          duration: Date.now() - start,
          error: String(error),
        };
      }
    },
  };
}

/**
 * 周期任务：记忆整理 + 深度清理
 */
function createCycleTask(workspaceDir: string): Task {
  return {
    name: 'memory_cleanup',
    layer: 'cycle',
    interval: 3600000,
    run: async () => {
      const start = Date.now();
      const memoryDir = path.join(workspaceDir, 'memory');
      const dailyDir = path.join(memoryDir, 'daily');
      const archiveDir = path.join(memoryDir, 'archive');
      
      let archived = 0;
      let cleaned = 0;
      
      if (!fs.existsSync(dailyDir)) {
        return { success: true, taskName: 'memory_cleanup', duration: 0 };
      }
      
      const now = Date.now();
      const ninetyDaysAgo = now - (90 * 24 * 60 * 60 * 1000);
      const thirtyDaysAgo = now - (30 * 24 * 60 * 60 * 1000);
      
      try {
        const files = fs.readdirSync(dailyDir).filter(f => f.endsWith('.jsonl'));
        
        for (const file of files) {
          const filePath = path.join(dailyDir, file);
          const stat = fs.statSync(filePath);
          
          // 90天以上 → 归档
          if (stat.mtimeMs < ninetyDaysAgo) {
            if (!fs.existsSync(archiveDir)) {
              fs.mkdirSync(archiveDir, { recursive: true });
            }
            fs.renameSync(filePath, path.join(archiveDir, file));
            archived++;
          }
        }
        
        return {
          success: true,
          taskName: 'memory_cleanup',
          duration: Date.now() - start,
          message: `Archived ${archived} files`,
        };
      } catch (error) {
        return {
          success: false,
          taskName: 'memory_cleanup',
          duration: Date.now() - start,
          error: String(error),
        };
      }
    },
  };
}

// ============================================================================
// 心跳系统主控制器
// ============================================================================

export class HeartbeatSystem {
  private config: HeartbeatConfig;
  private state: HeartbeatState;
  private running: boolean = false;
  private lastRun: Record<string, number> = {
    pulse: 0,
    rhythm: 0,
    cycle: 0,
  };
  
  private registry: TaskRegistry;
  private timer: NodeJS.Timeout | null = null;
  private listeners: ((result: TaskResult) => void)[] = [];

  constructor(workspaceDir?: string) {
    this.config = initConfig(workspaceDir);
    this.state = this.loadState();
    this.registry = new TaskRegistry();
    
    // 注册内置任务
    this.registerTask(createPulseTask());
    this.registerTask(createRhythmTask(this.config.workspaceDir));
    this.registerTask(createCycleTask(this.config.workspaceDir));
  }

  // ==================== 状态管理 ====================
  
  private loadState(): HeartbeatState {
    const defaultState: HeartbeatState = {
      startTime: new Date().toISOString(),
      lastPulse: null,
      lastRhythm: null,
      lastCycle: null,
      totalPulses: 0,
      totalRhythms: 0,
      totalCycles: 0,
      failedTasks: 0,
    };

    if (!fs.existsSync(this.config.stateFile)) {
      return defaultState;
    }

    try {
      const data = fs.readFileSync(this.config.stateFile, 'utf-8');
      return { ...defaultState, ...JSON.parse(data) };
    } catch {
      return defaultState;
    }
  }

  private saveState(): void {
    try {
      const dir = path.dirname(this.config.stateFile);
      if (!fs.existsSync(dir)) {
        fs.mkdirSync(dir, { recursive: true });
      }
      fs.writeFileSync(this.config.stateFile, JSON.stringify(this.state, null, 2));
    } catch (error) {
      console.error('[heartbeat] Failed to save state:', error);
    }
  }

  // ==================== 任务管理 ====================
  
  /**
   * 注册任务
   */
  registerTask(task: Task): void {
    this.registry.register(task);
    console.log(`[heartbeat] Registered task: ${task.name} (${task.layer})`);
  }
  
  /**
   * 注销任务
   */
  unregisterTask(name: string): void {
    this.registry.unregister(name);
  }
  
  /**
   * 启用/禁用任务
   */
  setTaskEnabled(name: string, enabled: boolean): void {
    const task = this.registry.getTasks().find(t => t.name === name);
    if (task) {
      task.enabled = enabled;
    }
  }
  
  /**
   * 获取所有任务统计
   */
  getTaskStats(): TaskStats[] {
    return this.registry.getAllStats();
  }
  
  /**
   * 获取单个任务统计
   */
  getTaskStat(name: string): TaskStats | undefined {
    return this.registry.getStats(name);
  }

  // ==================== 事件 ====================
  
  /**
   * 添加任务结果监听器
   */
  onTaskComplete(listener: (result: TaskResult) => void): void {
    this.listeners.push(listener);
  }
  
  /**
   * 移除监听器
   */
  offTaskComplete(listener: (result: TaskResult) => void): void {
    this.listeners = this.listeners.filter(l => l !== listener);
  }
  
  private notifyListeners(result: TaskResult): void {
    for (const listener of this.listeners) {
      try {
        listener(result);
      } catch (error) {
        console.error('[heartbeat] Listener error:', error);
      }
    }
  }

  // ==================== 心跳执行 ====================
  
  private updateLastRun(taskName: string, layer: string): void {
    this.lastRun[taskName] = Date.now();
    const now = new Date().toISOString();
    
    if (layer === 'pulse') {
      this.state.lastPulse = now;
      this.state.totalPulses++;
    } else if (layer === 'rhythm') {
      this.state.lastRhythm = now;
      this.state.totalRhythms++;
    } else if (layer === 'cycle') {
      this.state.lastCycle = now;
      this.state.totalCycles++;
    }
    
    this.saveState();
  }

  /**
   * 执行单次心跳
   */
  async tick(): Promise<void> {
    const elapsed = Date.now();

    // Pulse 层任务（每5秒）
    for (const task of this.registry.getTasks('pulse')) {
      if (!task.enabled) continue;
      if (elapsed - (this.lastRun[task.name] || 0) >= task.interval) {
        await this.runTask(task);
      }
    }

    // Rhythm 层任务（每60秒）
    for (const task of this.registry.getTasks('rhythm')) {
      if (!task.enabled) continue;
      if (elapsed - (this.lastRun[task.name] || 0) >= task.interval) {
        await this.runTask(task);
      }
    }

    // Cycle 层任务（每1小时）
    for (const task of this.registry.getTasks('cycle')) {
      if (!task.enabled) continue;
      if (elapsed - (this.lastRun[task.name] || 0) >= task.interval) {
        await this.runTask(task);
      }
    }
  }
  
  /**
   * 运行单个任务
   */
  private async runTask(task: Task): Promise<void> {
    const start = Date.now();
    
    try {
      const result = await task.run();
      result.duration = Date.now() - start;
      result.taskName = task.name;
      
      this.registry.updateStats(result, result.duration);
      this.updateLastRun(task.name, task.layer);
      
      if (!result.success) {
        this.state.failedTasks++;
        this.saveState();
      }
      
      this.notifyListeners(result);
      
      console.log(`[heartbeat] ${task.name}: ${result.success ? '✓' : '✗'} (${result.duration}ms)`);
    } catch (error) {
      const duration = Date.now() - start;
      const result: TaskResult = {
        success: false,
        taskName: task.name,
        duration: duration,
        error: String(error),
      };
      
      this.registry.updateStats(result, duration);
      this.updateLastRun(task.name, task.layer);
      this.state.failedTasks++;
      this.saveState();
      
      console.error(`[heartbeat] ${task.name}: ✗ Error:`, error);
    }
  }

  // ==================== 生命周期 ====================
  
  /**
   * 启动心跳循环
   */
  start(): void {
    if (this.running) return;
    
    this.running = true;
    console.log('[heartbeat] Heartbeat system started');
    console.log(`[heartbeat] Tasks: ${this.registry.getTasks().map(t => t.name).join(', ')}`);
    
    // 立即执行一次
    this.tick().catch(console.error);
    
    // 设置定时器
    this.timer = setInterval(() => {
      if (!this.running) return;
      this.tick().catch(console.error);
    }, this.config.pulseInterval);
  }

  /**
   * 停止心跳
   */
  stop(): void {
    this.running = false;
    if (this.timer) {
      clearInterval(this.timer);
      this.timer = null;
    }
    console.log('[heartbeat] Heartbeat system stopped');
  }

  /**
   * 立即触发一次任务
   */
  async triggerTask(name: string): Promise<TaskResult | null> {
    const task = this.registry.getTasks().find(t => t.name === name);
    if (!task) return null;
    
    await this.runTask(task);
    const stats = this.registry.getStats(name);
    return stats?.lastResult || null;
  }

  /**
   * 获取系统状态
   */
  getStatus(): {
    running: boolean;
    config: HeartbeatConfig;
    state: HeartbeatState;
    lastRun: Record<string, number>;
    tasks: TaskStats[];
  } {
    return {
      running: this.running,
      config: this.config,
      state: this.state,
      lastRun: this.lastRun,
      tasks: this.getTaskStats(),
    };
  }
}

// ============================================================================
// 导出
// ============================================================================

export function createHeartbeatSystem(workspaceDir?: string): HeartbeatSystem {
  return new HeartbeatSystem(workspaceDir);
}

export default HeartbeatSystem;
